### ISSUE TITLE
macOS 15 / Sequoia Installation Fails with "The installer could not install the software because there was no software found to install"

### ISSUE THREAD

**Issue by cvoisey** [Ylianst/MeshCentral#6398]
**Describe the bug**
When trying to install the agent on my Sequoia macs, I am getting the following error:
<img width="620" alt="image" src="https://github.com/user-attachments/assets/329d6d94-409e-4784-a66b-2d007f413c64">

This is an Intel mac and I have tried both installers (x86 & Universal).  Installs on other, older macOS from this server works fine.

 - OS: Ubuntu
 - Virtualization: N/A
 - Network: AWS EC2
 - Version: [e.g. 1.0.43]
 - Node: [e.g. 18.4.0]

**Comment by cvoisey** [comment]
I am already running on the latest:
![image](https://github.com/user-attachments/assets/9555314e-b3a6-4997-b436-385f5c8ee906)

**Comment by si458** [comment]
oh right ok, well ur bug report says 1.0.48,
is it ARM or intel you are using?
did you download the universal from the web ui OR the actual version for your device?

**Comment by cvoisey** [comment]
Oh, sorry, I started filling that all out and submitted it without finishing.  My bad.

I am on an Intel Mac, and I tried the Universal and the x86 versions.  Neither have worked.

**Comment by si458** [comment]
hmmm im guessing apple have change something!
as it works on sonoma intel macbook pro here!
might need to do some more testing :(

**Comment by si458** [comment]
just updating my mac mini arm from ventura to sonoma,
once thats done, i can upgrade xcode to 16,
then i can get UTM to run sequoia!
and if the arm doesnt work, then intel wont work

i dont want to upgrade my macbook pro sonoma to sequoia as its my daily driver and i dont want things to break!

**Comment by si458** [comment]
ok this might be a pain 
`Contents/Packages/internal.pkg not supported by PK`
`Unable to use PK session due to incompatible packages. Terminating.`
full logs attached from my arm 15 vm but its the exact same issue 👎 
[Installer_Log_20-Sep-2024.txt](https://github.com/user-attachments/files/17081155/Installer_Log_20-Sep-2024.txt)

**Comment by cvoisey** [comment]
I wish I could give you a more detailed answer on this, but it sounds similar to something one of my developers recently encountered.  I'll talk to them on Monday, but I don't think they resolved it yet.  They were checking the stricter notarizations that are required with macOS Sequoia.

**Comment by si458** [comment]
@cvoisey notarizations isnt required (just really really really recommended!)

i have been able to use an app called `Packages` which generates PKGs (both flat and bundle)
and get it to copy over the meshagent and enable startup etc with no problems!

its something with the way the .pkg is generated/built that im guessing Sequoia just doesnt like anymore

sadly the app `Packages` makes the PKGs in a different way for me so i cant just extract it, then build a zip 
for windows/linux machines to then rebuild!

**Comment by cvoisey** [comment]
Are you doing the current release in bundle format?  What about trying to do it in flat?  That should work better with newer macOS

**Comment by si458** [comment]
@cvoisey ive tried both, flat and bundle, both types from the `packages` app work a treat!
i just cant explore/extract the pkg and thats the issue

its because its saving it as`XAR archive` format
```
simon@simonlappy build % file MeshAgent.pkg 
MeshAgent.pkg: xar archive compressed TOC: 867, SHA-1 checksum, contains  zlib compressed data
```

but the old .pkg we have create returns it as a directory?
```
simon@simonlappy ~ % file /Users/simon/Downloads/myremoteagent\ 2.mpkg/Contents/Packages/internal.pkg
/Users/simon/Downloads/myremoteagent 2.mpkg/Contents/Packages/internal.pkg: directory
```
so unless im doing something wrong, this could be why `Sequoia` no longer works as they need it in `xar` format

**Comment by cvoisey** [comment]
I've been trying to figure out something as well and have been coming to similar conclusions.  I have to put it down for now, but I have a few more ideas I'll try to get to tomorrow.   I'd love it if I could contribute.  You've done a great job with this project.

**Comment by si458** [comment]
mad idea 💡 
why dont we create a DMG, include the meshagent as binary format,
BUT we hide it, and just display a script file
and the background is like
```
DOUBLE CLICK THIS SCRIPT FILE BELOW TO INSTALL
PLEASE NOTE, YOU WILL BE ASKED FOR YOUR PASSWORD IN A BLACK POP UP WINDOW
```

**Comment by cvoisey** [comment]
From a short term perspective, to get it installed on newer OSs, I could see that.  From a longer term though, I'm thinking about the UX.  

Personally, I want to still dive into the whole build experience in my hosting and make sure I sign everything and ensure it is trusted, as that will look better to my customers.

**Comment by si458** [comment]
@cvoisey fully agree
im having problems now creating the dmgs on linux and most importantly windows!
why apple has to make everything difficult i will never understand

**Comment by carlosndiaye** [comment]
Hello. Thank you very much for your work. I have the same problem. I managed to install it with the binary on the terminal but I get a black screen when I try to connect in Remote Desktop. I can't access certain directories via the shell either.

**Comment by si458** [comment]
@carlosndiaye before u connect, u need to go into settings the privacy and security, then screen recording and allow the application 

Also, go into accessibility and allow ur app there too

As for files, u need to allow ur app inside the full disk access

All options are inside the privacy & security

**Comment by carlosndiaye** [comment]
Thanks. Apparently, when I allow the settings as mentioned, the desktop is displayed and you can interact with it. However, when I reconnect, the black screen comes back even though the permissions are still active. What could be the cause of this?

**Comment by si458** [comment]
@carlosndiaye I'm not sure? That seems weird?
Can u open a separate issue please for that?
Follow the bug report too plz 🙏 🙂

**Comment by jakobzudrell** [comment]
Here to confirm that the installation via binary worked fine on Sequoia (ARM).
I'll try to wrap the binary installer within a DMG, however I am new to this stuff so don't expect something fancy or anything at all ;) I'll try my luck and post back if I have something..

Just for completeness, here my output from the mpkg install. But I think that's no new information.
`
installer: Package name is Mesh Agent
Oct  2 09:22:48  installer[33094] <Debug>: Product archive /Users/user/Downloads/meshagent/MeshAgent.mpkg trustLevel=100
Oct  2 09:22:48  installer[33094] <Error>: PKInstallRequest: failed to initialize. Error: Error Domain=PKInstallRequestErrorDomain Code=2 "Specifier Description:<PKPackageSpecifier>:
	{
	    URL = "file:///Users/user/Downloads/meshagent/MeshAgent.mpkg/Contents/Packages/internal.pkg";
	    identifier = "internal.pkg";
	    options = 0;
	    version = "1.0";
	}" UserInfo={NSLocalizedFailureReason=Specifier Description:<PKPackageSpecifier>:
	{
	    URL = "file:///Users/user/Downloads/meshagent/MeshAgent.mpkg/Contents/Packages/internal.pkg";
	    identifier = "internal.pkg";
	    options = 0;
	    version = "1.0";
	}}
Oct  2 09:22:48  installer[33094] <Debug>: External component packages (1) trustLevel=100 (trust evaluation failed)
installer: Upgrading at base path /
installer: Preparing for installation….....
Oct  2 09:22:48  installer[33094] <Debug>: -[IFDInstallController(Private) _buildInstallPlanReturningError:]: location = file://localhost
Oct  2 09:22:48  installer[33094] <Debug>: -[IFDInstallController(Private) _buildInstallPlanReturningError:]: file://localhost/Users/user/Downloads/meshagent/MeshAgent.mpkg/Contents/Packages/internal.pkg
Oct  2 09:22:48  installer[33094] <Info>: Set authorization level to root for session
Oct  2 09:22:48  installer[33094] <Info>: Authorization is being checked, waiting until authorization arrives.
Oct  2 09:22:48  installer[33094] <Info>: Administrator authorization granted.
Oct  2 09:22:48  installer[33094] <Info>: Packages have been authorized for installation.
2024-10-02 09:22:48.284 installer[33094:3497317] Package /Users/user/Downloads/meshagent/MeshAgent.mpkg/Contents/Packages/internal.pkg uses a deprecated pre-10.2 format (or uses a newer format but is invalid).
Oct  2 09:22:48  installer[33094] <Debug>: Package file://localhost/Users/user/Downloads/meshagent/MeshAgent.mpkg/Contents/Packages/internal.pkg not supported by PK
Oct  2 09:22:48  installer[33094] <Debug>: Unable to use PK session due to incompatible packages. Terminating.
Oct  2 09:22:48  installer[33094] <Error>: Install failed: The Installer could not install the software because there was no software found to install.
installer: The upgrade failed. (The Installer could not install the software because there was no software found to install. The Installer could not install the software.)
`

**Comment by appatterson** [comment]
Running into this same issue. How are you using Packages to reformat the .pkg?

**Comment by si458** [comment]
@appatterson it was very difficult

I had to basically add all the files in and msh and then tell it to run the script from inside the mpkg to place them where they need to be

Ill show screenshots 2moz of it :)

**Comment by appatterson** [comment]
@si458 that would be much appreciated, thanks!

**Comment by si458** [comment]
@appatterson hope these screenshots help!
basically create the folder where stuff goes in payload,
then add all the files from inside the pkg thats inside the mpkg from meshcentral
then set the payload file to the payload.sh file but you need to modify the file slightly!
<img width="1792" alt="Screenshot 2024-10-05 at 00 15 39" src="https://github.com/user-attachments/assets/bbbdddd9-d27c-4b10-b995-a2265fb86adb">
<img width="1792" alt="Screenshot 2024-10-05 at 00 14 42" src="https://github.com/user-attachments/assets/5a9d5af3-ab71-4958-b101-4ed2722ea4ec">
<img width="1792" alt="Screenshot 2024-10-05 at 00 15 16" src="https://github.com/user-attachments/assets/d678befb-0b6b-4066-9da6-f5e425e7abf1">
<img width="1792" alt="Screenshot 2024-10-05 at 00 15 45" src="https://github.com/user-attachments/assets/dc8fe720-b7fe-4fa0-be79-caef5c99d978">
<img width="998" alt="Screenshot 2024-10-05 at 00 16 09" src="https://github.com/user-attachments/assets/11324c2d-66b7-43b3-8ef9-96d8f658e601">

**Comment by DaanSelen** [comment]
> @appatterson hope these screenshots help! basically create the folder where stuff goes in payload, then add all the files from inside the pkg thats inside the mpkg from meshcentral then set the payload file to the payload.sh file but you need to modify the file slightly! <img alt="Screenshot 2024-10-05 at 00 15 39" width="1792" src="https://private-user-images.githubusercontent.com/765314/373826997-bbbdddd9-d27c-4b10-b995-a2265fb86adb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2ODM1NTIsIm5iZiI6MTcyODY4MzI1MiwicGF0aCI6Ii83NjUzMTQvMzczODI2OTk3LWJiYmRkZGQ5LWQyN2MtNGIxMC1iOTk1LWEyMjY1ZmI4NmFkYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxMVQyMTQ3MzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iNzgyNmU4YmQyN2Y0MjlmYjg3YThiNWMwNzJiNzFmMWQ3ZWFiNTUzMzU4ZDhkOWYyZmViODZiZmUzZjUyYThiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.vb1ZB2tH8T61LyfwXjLdEmTZ5xAoScm-tguAh3Xrums"> <img alt="Screenshot 2024-10-05 at 00 14 42" width="1792" src="https://private-user-images.githubusercontent.com/765314/373827001-5a9d5af3-ab71-4958-b101-4ed2722ea4ec.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2ODM1NTIsIm5iZiI6MTcyODY4MzI1MiwicGF0aCI6Ii83NjUzMTQvMzczODI3MDAxLTVhOWQ1YWYzLWFiNzEtNDk1OC1iMTAxLTRlZDI3MjJlYTRlYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxMVQyMTQ3MzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mYzM1YzgxYzkxYWRmYTUyMTRhNDdlNWE2MWYzNzEyZmRlYmNhODI5OGQ4MDIxZWI1NmZiZGUzN2RmODY4YzNmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.n_d_0Vyf3IZB1BzSjUssE3zZmNuo9E_aRXhGScopwO4"> <img alt="Screenshot 2024-10-05 at 00 15 16" width="1792" src="https://private-user-images.githubusercontent.com/765314/373827010-d678befb-0b6b-4066-9da6-f5e425e7abf1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2ODM1NTIsIm5iZiI6MTcyODY4MzI1MiwicGF0aCI6Ii83NjUzMTQvMzczODI3MDEwLWQ2NzhiZWZiLTBiNmItNDA2Ni05ZGE2LWY1ZTQyNWU3YWJmMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxMVQyMTQ3MzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wNzE4YTI0Y2RlZDVjNTY1YWJlOGMzMmZlMmI2Njk1MDExYjIzY2M2MThlZGFmNDI3MzBlZDViZDFmNWRjYmM1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.trBCEGw2htNl5tpNoC2JpSEpuIVh3NprGalCBNSkmGQ"> <img alt="Screenshot 2024-10-05 at 00 15 45" width="1792" src="https://private-user-images.githubusercontent.com/765314/373827011-dc8fe720-b7fe-4fa0-be79-caef5c99d978.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2ODM1NTIsIm5iZiI6MTcyODY4MzI1MiwicGF0aCI6Ii83NjUzMTQvMzczODI3MDExLWRjOGZlNzIwLWI3ZmUtNGZhMC1iZTc5LWNhZWY1Yzk5ZDk3OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxMVQyMTQ3MzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lNDhlYzQ2N2M2YzBhZmE1NTY1NjQ2ODI3NmUyYjM3ZmY3ZDJhMDY4MWY4ZjI4NWYwYmNiMDVkZjA1NTQyMzRjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Z6L8jCSXCr4KGdeT7Y4Z_xHZvpdDdEhkWzU7m5E2Ig0"> <img alt="Screenshot 2024-10-05 at 00 16 09" width="998" src="https://private-user-images.githubusercontent.com/765314/373827017-11324c2d-66b7-43b3-8ef9-96d8f658e601.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2ODM1NTIsIm5iZiI6MTcyODY4MzI1MiwicGF0aCI6Ii83NjUzMTQvMzczODI3MDE3LTExMzI0YzJkLTY2YjctNDNiMy04ZWY5LTk2ZDhmNjU4ZTYwMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxMVQyMTQ3MzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hZjk2NzE4NGQwZTc1OTRmMmMwYTZlYjJjZGQ1MmNiYmNiY2I3NGFlNjE3MzEyY2QzYzQwZTBhZGEzNmQyZmRkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.jH0Zq94HFqSEOdpifPZDgdVAVXRyaZd2-2le_KoTWjM">

This is a fix?

**Comment by johnnyq** [comment]
@si458 Maybe the new MacOS 15.1 Update coming out later this month will fix the issue, fingers crossed.
I have held my Mac clients back from updating to 15.0

**Comment by esgheorghe** [comment]
Hello,

any news about MacOS Agent in Sequoia ?

**Comment by si458** [comment]
@esgheorghe nope :(

I've had no luck no matter what I try, but i don't understand the c++ code fully so the isn't much i can do without help from someone who does

**Comment by ramphex** [comment]
Same issue here on Sequoia 15.1

**Comment by jakobzudrell** [comment]
@si458 could you link me to the place where this ominous C++ code resides? :P
 
Sadly I am by no means an expert on C++, but I'd be really interested in taking a look at it, playing around a little and maybe even be able to contribute back a little at some point...

**Comment by si458** [comment]
@jakobzudrell check the meshagent repo!
All the codes in there https://github.com/Ylianst/MeshAgent

**Comment by ramphex** [comment]
Any update on this? Current I'm seeing issues trying to install on MacOS and also issues trying to install on Windows. Seems like the client is falling behind with two major OS's updates.

**Comment by si458** [comment]
@ramphex sadly, without a Mac developer who can identify why we can't get remote control to work, it won't be possible

As for Windows 11 24h2, it's already been fixed, new agents have been tested and work, just need new release being done,
or
In the meantime, u simply install the wmic add-on and then install meshagent, and it carries on like nothing happened

So we arent really falling behind at all...

We are an open-source application and open source developers without any physical funding so things take time

**Comment by ramphex** [comment]
> @ramphex sadly, without a Mac developer who can identify why we can't get remote control to work, it won't be possible
> 
> As for Windows 11 24h2, it's already been fixed, new agents have been tested and work, just need new release being done, or In the meantime, u simply install the wmic add-on and then install meshagent, and it carries on like nothing happened
> 
> So we arent really falling behind at all...
> 
> We are an open-source application and open source developers without any physical funding so things take time

I get it. I appreciate the software and the time you guys put into it.

**Comment by DaanSelen** [comment]
Linux was not interrupted?

**Comment by johnnyq** [comment]
@DaanSelen Yes it is if you use a Wayland Distro (Latest versions of Ubuntu, Cosmic and Fedora)
Mint is unaffected but will soon be once they switch fully to wayland
but this is totally unrelated to this thread and very different issue

**Comment by eried** [comment]
No way yet to install on latest MacOS right? 🤓 could I install the Linux version on a mac?

**Comment by si458** [comment]
@eried it works for macos 14 or below, just not 15 or above!
We are still looking for a developer to help replace the old deprecated functions with screencapturekit!
And no sadly as they are compiled differently so it won't work

**Comment by eried** [comment]
oh, yeah we had a machine that got updated and now it does not work. I though it was just permissions, but I see is not. I have never used macos, so I cant really help :( (this machine is only for compiling stuff for the appstore, but nobody uses it eitherway)

**Comment by elpibedeoro** [comment]
Hi,

Have the same problem in graphical mode but I was able to install it in command line mode.
Depending on your CPU architecture, get the meshagent binary through the wget command (brew install wget). 

![meshscreenshot](https://github.com/user-attachments/assets/188a20a7-f831-4bb8-b0dd-435dc54b7846)

Chmod the binary and install it (./meshagent -install). 

After have set the security settings, everything works but at the next boot the connection is lost... this can be fixed by renaming the daemon, as as we have already seen in the past.
1.	sudo sed -i '' '/<key>Label<\/key>/,/<\/string>/s/meshagent/meshdaemon/'  /Library/LaunchDaemons/meshagent.plist
2.	sudo mv /Library/LaunchDaemons/meshagent.plist /Library/LaunchDaemons/meshdaemon.plist

The connection is ok this time but the Desktop view show me a black screen... any idea ?

PS:  Hope this won't be off topic but if it can help someone...

**Comment by si458** [comment]
@elpibedeoro the black screen bug is unfortunately a HUGE bug when using macos 15 or above,
it hasnt been fixed yet due to apple removing the required KVM module we use for the screen recording :(
the patch/fix is to replace the KVM module with screencapturekit but thats a big job! and we are still looking for a developer to look into it for us
the is a whole issue about the bug here https://github.com/Ylianst/MeshCentral/issues/6402

**Comment by SinisterSpatula** [comment]
I installed the latest meshcentral agent on Mac OS Sequoia 15.3.1 and it's working for me.  I just use the web-vnc option for remote control, to get around the black screen issue and it's working well that way for me.  I had to faff about with the wget, and editing the launcher plist so it auto-starts on reboot, but other than that it seems to be good.

**Comment by si458** [comment]
@SinisterSpatula thats good to know! 
can you explain in the issue https://github.com/Ylianst/MeshCentral/issues/6402 HOW you fixed the launcher plist?
i think i have fixed it here in the meshagent code (so macs will need a new agent being built) but wanted to be 110% sure!
i still havent fixed the remote desktop issue but thats a LONG game at the moment

**Comment by DaanSelen** [comment]
Maybe a weird idea but also for Linux wouldnt it be possible to ocnnect meshcentral to a vnc server on an agent and get screen that way? Then we upstream things! Xfreerdp has wayland support. Cant we connect like meshcentral agent to display that?

**Comment by si458** [comment]
@DaanSelen yes u can already use the web-vnc with linux, just setup a vnc server then use the option
i use this from time to time on my raspberry pi if the remote isnt working 👍

**Comment by chdodier** [comment]
@si458 @Ylianst   Premièrement Merci pour votre travail!

I managed to install meshagent Mac OS Sequoia 15.3.1 using version 1.1.38 of the server but with the new agent 

![Image](https://github.com/user-attachments/assets/a4c1c0a9-b786-4233-a207-56209ad553e6)

I change my default terminal first to zsh  

then downloaded meshagent into my download folder

and ran this : chmod +x /Users/%Myusername%/Downloads/meshagent
then : sudo /Users/%Myusername%/Downloads/meshagent -install

then os blocks it : you need to go into system setting / security  and unblock it

Then still in security : screen recording and audio : Allow meshagent  you should probably allow disk acces also

Unfortunatly,  the video worked only once, when i disconnected it, I could not reconnect it
Will try again later

**Comment by si458** [comment]
@chdodier we dont have a new agent for mac os, so ur using the normal 2022 agent
and yes thats whats happening at the moment, u get screen recording ONCE and ONLY ONCE, AND only after everything u rinstall the agent
its very strange

**Comment by quizbrain** [comment]
I'm using a DWservice client which works great, in the meantime, for screen capture.
Just saying in case someone needs a free solution in the mean time.
Thanks.

**Comment by rpicat** [comment]
Hello,

I'm having the same problem on a Mac running version 15.3. If I understand correctly, even if I work around the problem using command lines, remote access doesn't work.

Will there be a fix for the agent?

Thank you very much.

**Comment by MonsterITServices** [comment]
Any luck on coding for 15.4 and above?

**Comment by vikujangid** [comment]
Is there anyone who can resolve this installation issue on version 15.4 and above (using UI Installer)?

**Comment by eried** [comment]
> Is there anyone who can resolve this installation issue on version 15.4 and above (using UI Installer)?

probably not, I think the "mac" part is always a nuisance on companies (but needed due iphones, appstore, blabla) 😂 so even worse if needs contributions to work here

**Comment by Junosprite007** [comment]
Dang, it's a bummer when such a great program like this has a pesky bug from some big tech company that wants to update there stuff way too often. Apple is so annoying sometimes. I guess I'll just say thank you for all the work that you've done on this. Still works great on Windows! I can fix my parents' computer without having to walk them through it!

I might attempt to fix this on my Mac M1 machine, but I too am not very well versed on macOS. I'll just use Claude or something haha.

**Comment by eried** [comment]
> I might attempt to fix this on my Mac M1 machine, but I too am not very well verses on macOS. I'll just use Claude or something haha.

Go for it, with claude you will probably get somewhere and help everyone here

**Comment by Junosprite007** [comment]
OK, I'm still not going to attempt to fix this myself yet, but I did prompt Claude Sonnet 4.0 to do some deep research on the topic to try and help find a solution. Below is my prompt along with its solution after it finished its research. This is just incase someone wants to try figuring it out first. Both the following prompt and follow-up prompt were executed with Claude's Research BETA option activated. And incase it's more helpful, here's a couple sentence summary after the research was complete:

**Problem**: MeshCentral generates macOS installers using ZIP-based bundle-style .pkg format that macOS Sequoia no longer accepts, causing "no software found to install" errors.

**Fix**: Replace the ZIP library (yazl/yauzl) package generation in webserver.js with native macOS pkgbuild commands to create flat XAR-format packages compatible with Sequoia.

## Prompt

I'm trying to help the open-source community by fixing a bug for the software called "MeshCentral", located at the public git rep: https://github.com/Ylianst/MeshCentral

The specific bug is noted be the issues located here: https://github.com/Ylianst/MeshCentral/issues/6398

I'll paste the text of this issue as seen on GitHub as as well as as an image that a user included. This image is the same thing that I received when I tried to install a mesh agent on my macOS M1 (Apple Silicone) machine.

Please keep in mind that I am having my own personal issue on a Mac M1 system, whereas this logged bug is occurring on an Intel Mac. This is just for clarification to help you find what the issue is. I have both repositories for MeshCentral and MeshAgent already downloaded onto my local environment so I can navigate through the files and directories as necessary. Please tell me where I can search in order to find the bug that is causing this, and finally helped me fix this so that I can create a poll request to the repo.

Please go as in depth as possible to find the possible cause of this issue. I want to modify as few parts of the official code as possible in order to get this working again. Also, please let me know which programming languages. I will need to be using in order to do this. Good luck with your extremely in depth research. I appreciate it!

## Claude Response
Strategized immediate research for MeshCentral installer bug resolution
7s

This is a fascinating and complex macOS installer issue that's been plaguing the MeshCentral community! I can see from the GitHub issue that this involves Apple's changing installer requirements in Sequoia, package format issues, and the transition from bundle-style to XAR archive formats. Let me dive deep into researching the technical root cause and potential solutions.


MeshCentral macOS installer bug

Research complete
•
354 sources and counting...
371 sources
•
5m 19s


Your MeshCentral macOS Sequoia bug fix analysis is ready. It explains why the installer fails on macOS 15 (Sequoia), identifies the root cause in the bundle-style package format, and provides both immediate and complete solutions. The report includes code samples for implementing flat XAR packages with proper signing, required environment setup instructions, testing procedures, and a development roadmap for full compatibility with Apple's stricter security requirements.


MeshCentral macOS Sequoia Installer Bug Fix

Document:

#############################################################################################

# MeshCentral macOS Sequoia Installer Bug Fix

macOS Sequoia's enhanced installer validation has broken MeshCentral's .pkg generation due to format incompatibility. The current bundle-style package format is rejected by Sequoia's stricter security requirements, requiring a shift to modern flat package format with proper code signing and notarization.

## Root cause analysis

**Sequoia validation changes**: macOS 15.x introduced stricter installer validation that effectively deprecates bundle-style .pkg packages in favor of flat XAR-based packages. The error "The installer could not install the software because there was no software found to install" occurs when Sequoia's enhanced installer daemon (installd) encounters older bundle-style packages that lack proper XAR archive structure and digital signatures.

**MeshCentral's current approach**: The package generation code in `webserver.js` creates traditional **bundle format** .mpkg files using Node.js ZIP libraries (`yauzl` and `yazl`). This creates directory-based packages that Sequoia's validation engine cannot process properly. The current structure generates:

```
MeshAgent.mpkg/
├── Contents/
│   ├── distribution.dist
│   └── Packages/
│       └── internal.pkg/
│           └── Contents/
│               ├── meshagent_osx64.bin
│               └── meshagent_osx64.msh
```

**Technical incompatibility**: Bundle packages cannot be digitally signed for integrity verification and lack the XAR archive structure with proper table of contents that Sequoia requires. The current ZIP-based approach doesn't generate proper XAR-format flat packages that modern macOS expects.

## Code location and modification requirements

**Primary files requiring changes**:
- `webserver.js` - Contains `handleMeshOsxAgentRequest()` function (lines ~8000-8500)
- `package.json` - Dependencies need updating to support proper package generation

**Current problematic code**: The `handleMeshOsxAgentRequest()` function uses `yazl` library to create ZIP archives packaged as .mpkg bundles. This entire approach must be replaced with native macOS packaging tools.

## Minimal fix implementation

**Phase 1: Immediate compatibility fix**

Replace the ZIP-based package generation with native macOS tools called via Node.js `child_process`. Update the `handleMeshOsxAgentRequest()` function to:

1. **Create flat package format**: Use `pkgbuild` instead of ZIP libraries
2. **Generate proper package structure**: Create XAR archive with correct metadata
3. **Enable basic signing**: Add certificate-based signing for validation

**Code changes required**:

```javascript
// Replace this section in webserver.js handleMeshOsxAgentRequest()
const { spawn } = require('child_process');
const path = require('path');
const fs = require('fs');

// Instead of yazl/yauzl ZIP creation, use native macOS tools
function createMacOSPackage(agentPath, configPath, outputPath, callback) {
    // Create component package using pkgbuild
    const pkgbuildArgs = [
        '--identifier', 'com.meshcentral.agent',
        '--version', '1.0.0',
        '--root', agentPath,
        '--install-location', '/usr/local/mesh',
        '--timestamp',
        outputPath
    ];
    
    // Add signing if certificate available
    if (process.env.MESH_MACOS_CERT) {
        pkgbuildArgs.push('--sign', process.env.MESH_MACOS_CERT);
    }
    
    const pkgbuild = spawn('pkgbuild', pkgbuildArgs);
    
    pkgbuild.on('close', (code) => {
        if (code === 0) {
            callback(null, outputPath);
        } else {
            callback(new Error(`Package creation failed with code ${code}`));
        }
    });
}
```

**Distribution XML template**: Create proper installer metadata:

```xml
<?xml version="1.0" encoding="utf-8"?>
<installer-gui-script minSpecVersion="2">
    <title>MeshCentral Agent</title>
    <organization>com.meshcentral</organization>
    <domains enable_localSystem="true"/>
    <options customize="never" require-scripts="false"/>
    <pkg-ref id="com.meshcentral.agent"/>
    <choices-outline>
        <line choice="default">
            <line choice="com.meshcentral.agent"/>
        </line>
    </choices-outline>
    <choice id="default"/>
    <choice id="com.meshcentral.agent" visible="false">
        <pkg-ref id="com.meshcentral.agent"/>
    </choice>
    <pkg-ref id="com.meshcentral.agent" version="1.0" onConclusion="none">meshagent-component.pkg</pkg-ref>
</installer-gui-script>
```

## Complete solution implementation

**Phase 2: Full compatibility with signing and notarization**

For production-ready fix, implement comprehensive solution:

1. **Code signing setup**: Obtain Developer ID certificates from Apple
2. **Notarization integration**: Add automated notarization to build process
3. **Multi-architecture support**: Ensure Universal binary compatibility

**Enhanced package generation function**:

```javascript
async function createSignedMacOSPackage(agentBinary, configFile, outputPath) {
    const tempDir = path.join(os.tmpdir(), 'meshcentral-pkg-' + Date.now());
    fs.mkdirSync(tempDir, { recursive: true });
    
    try {
        // 1. Prepare payload structure
        const payloadDir = path.join(tempDir, 'payload');
        const meshDir = path.join(payloadDir, 'usr', 'local', 'mesh');
        fs.mkdirSync(meshDir, { recursive: true });
        
        // Copy agent binary and config
        fs.copyFileSync(agentBinary, path.join(meshDir, 'meshagent'));
        fs.copyFileSync(configFile, path.join(meshDir, 'meshagent.msh'));
        
        // 2. Set proper permissions
        fs.chmodSync(path.join(meshDir, 'meshagent'), 0o755);
        
        // 3. Create component package
        const componentPkg = path.join(tempDir, 'meshagent-component.pkg');
        await execPromise('pkgbuild', [
            '--identifier', 'com.meshcentral.agent',
            '--version', '1.0.0',
            '--root', payloadDir,
            '--install-location', '/',
            '--timestamp',
            ...(process.env.MESH_INSTALLER_CERT ? ['--sign', process.env.MESH_INSTALLER_CERT] : []),
            componentPkg
        ]);
        
        // 4. Create distribution package if needed
        if (process.env.MESH_INSTALLER_CERT) {
            await execPromise('productbuild', [
                '--distribution', path.join(__dirname, 'distribution.xml'),
                '--package-path', tempDir,
                '--sign', process.env.MESH_INSTALLER_CERT,
                '--timestamp',
                outputPath
            ]);
        } else {
            fs.copyFileSync(componentPkg, outputPath);
        }
        
        return outputPath;
    } finally {
        // Cleanup
        fs.rmSync(tempDir, { recursive: true, force: true });
    }
}
```

## Required environment setup

**Development certificates**: For full compatibility, obtain Apple Developer certificates:
- Developer ID Application (for signing executables)
- Developer ID Installer (for signing packages)
- Apple Developer Program membership ($99/year)

**Environment variables**:
```bash
export MESH_INSTALLER_CERT="Developer ID Installer: Your Name (TEAMID)"
export MESH_APPLICATION_CERT="Developer ID Application: Your Name (TEAMID)"
export MESH_APPLE_ID="developer@example.com"
export MESH_TEAM_ID="TEAMID123"
```

**Build script integration**: Add to package.json:
```json
{
  "scripts": {
    "build:macos": "node scripts/build-macos-installer.js",
    "sign:macos": "codesign --sign \"$MESH_APPLICATION_CERT\" --options runtime --timestamp meshagent"
  }
}
```

## Testing and validation

**Pre-deployment validation**:
```bash
# Verify package structure
pkgutil --expand MeshAgent.pkg expanded/
tree expanded/

# Test installation
sudo installer -pkg MeshAgent.pkg -target /

# Validate signatures
spctl --assess --verbose --type install MeshAgent.pkg
```

**Sequoia-specific testing**:
- Test on clean macOS 15.x systems
- Verify Gatekeeper acceptance
- Validate post-installation agent functionality
- Test with different Sequoia versions (15.0, 15.1, 15.2+)

## Alternative workaround solution

For immediate deployment without certificates, implement **binary installer approach**:

```javascript
// Alternative: Create shell script installer instead of .pkg
function createShellInstaller(agentBinary, configFile, outputPath) {
    const installerScript = `#!/bin/bash
set -e

echo "Installing MeshCentral Agent..."

# Create destination directory
sudo mkdir -p /usr/local/mesh

# Copy files
sudo cp "${agentBinary}" /usr/local/mesh/meshagent
sudo cp "${configFile}" /usr/local/mesh/meshagent.msh

# Set permissions
sudo chmod 755 /usr/local/mesh/meshagent

# Create launch daemon
sudo cat > /Library/LaunchDaemons/com.meshcentral.agent.plist << 'EOF'
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>Label</key>
    <string>com.meshcentral.agent</string>
    <key>ProgramArguments</key>
    <array>
        <string>/usr/local/mesh/meshagent</string>
    </array>
    <key>RunAtLoad</key>
    <true/>
    <key>KeepAlive</key>
    <true/>
</dict>
</plist>
EOF

# Load launch daemon
sudo launchctl load /Library/LaunchDaemons/com.meshcentral.agent.plist

echo "MeshCentral Agent installed successfully!"
`;
    
    fs.writeFileSync(outputPath, installerScript);
    fs.chmodSync(outputPath, 0o755);
}
```

## Pull request implementation plan

**Immediate fix (PR #1)**:
1. Replace `yazl`/`yauzl` ZIP-based packaging with `pkgbuild` calls
2. Update `handleMeshOsxAgentRequest()` to use native macOS tools
3. Add proper error handling for systems without Xcode Command Line Tools
4. Maintain backward compatibility with older macOS versions

**Enhanced fix (PR #2)**:
1. Add certificate-based signing support
2. Implement notarization workflow
3. Add comprehensive testing for Sequoia compatibility
4. Update documentation for macOS deployment

**Files to modify**:
- `webserver.js` (primary changes)
- `package.json` (remove ZIP dependencies)
- `README.md` (update macOS requirements)
- `.github/workflows/` (add macOS build testing)

## Conclusion

The MeshCentral macOS Sequoia installer issue requires **fundamental changes** to package generation methodology rather than simple patches. The minimal fix involves replacing ZIP-based bundle packages with native macOS flat packages using `pkgbuild`. While this requires more complex implementation, it ensures long-term compatibility with Apple's security requirements and provides a foundation for enhanced features like proper code signing and notarization.

The immediate priority should be implementing the basic flat package generation to restore Sequoia compatibility, followed by the enhanced solution with proper signing for production deployments.

#############################################################################################

## Follow-up Claude prompt

Thank you. Could you do further research just to insure you did not leave out any crucial information? You don't need to restate anything or rewrite your document, but I would like you to investigate any potential oversights you or anyone appearing in your research may have had concerning this bug. Your previous document may already be complete, but if it's not, please correct any mistakes and add any further clarifications to a new, follow-up document. Any follow-up documents you create should not replace your existing documentation, but instead, it should only correct any errors or oversights and add new information.

Probed previous analysis for potential oversights and errors.
2s

I'll conduct a thorough follow-up investigation to identify any potential oversights, technical gaps, or missed solutions in my analysis of the MeshCentral macOS Sequoia installer bug.

MeshCentral macOS installer analysis

Research complete
•
Writing and citing report...
394 sources
•
4m 27s

Your MeshCentral macOS Sequoia installer bug analysis is ready. It confirms the root issue with incompatible package formats but identifies overlooked complexities: underlying ZIP corruption problems, significant implementation barriers for the pkgbuild solution, and simpler alternatives. The report details enterprise validation gaps, working interim solutions, and recommends a phased approach starting with enhanced binary installation documentation rather than immediate pkgbuild implementation.

Comprehensive QA Review: MeshCentral macOS Sequoia Installer Bug Analysis

Document:

#############################################################################################

# Comprehensive QA Review: MeshCentral macOS Sequoia Installer Bug Analysis

The previous technical analysis of MeshCentral's macOS Sequoia installer bug (#6398) was **fundamentally accurate** in identifying the root cause but missed several critical implementation details, alternative solutions, and validation gaps that could significantly impact the proposed fix.

## Validation of Original Analysis

The original analysis correctly identified that **MeshCentral's ZIP-based bundle-style .pkg generation is incompatible with Sequoia's flat XAR-format package requirements**. This conclusion is validated by Apple's official discontinuation of bundle-style package support in macOS Sequoia - a breaking change that affects all legacy installers using the older format.

However, the analysis revealed **deeper technical complexities and alternative approaches** that warrant significant consideration before implementing the proposed pkgbuild solution.

## Critical Technical Oversights Identified

### Persistent ZIP corruption amplifies the core problem

The investigation revealed that MeshCentral has **chronic ZIP archive corruption issues** producing consistent CRC errors in generated packages. Specifically, `MeshAgent.mpkg/Contents/distribution.dist` shows bad CRC 17c652b7 instead of the expected 00000000. This corruption existed before Sequoia but becomes fatal under Sequoia's stricter validation system.

This finding suggests the problem is **compound**: both format incompatibility AND data corruption. Simply switching to pkgbuild may not resolve the underlying corruption issues if the same data generation process is used.

### Implementation barriers significantly exceed initial assessment

The proposed pkgbuild solution has **substantial hidden requirements** not adequately addressed in the original analysis:

**Developer Infrastructure Requirements**: pkgbuild requires Xcode Command Line Tools and proper code signing. Most critically, Sequoia's enhanced security policies make **notarization mandatory** for seamless installation, requiring an Apple Developer Program membership ($99/year) and automated notarization infrastructure.

**Build Process Complexity**: The solution necessitates fundamental changes to MeshCentral's current build system, moving from simple ZIP file generation to a complex pipeline involving native macOS tools, certificate management, and Apple's notarization service.

**Platform Dependency**: pkgbuild only works on macOS hosts, creating deployment challenges for MeshCentral servers running on Linux or Windows - a significant portion of the user base.

## Alternative Solutions With Lower Implementation Costs

Research identified several **simpler approaches** that could achieve Sequoia compatibility without the complexity of switching to pkgbuild:

### Flat package conversion approach
Using `pkgutil --flatten` to convert existing bundle-style packages to flat format addresses the core compatibility issue with **minimal code changes**. This approach maintains current build processes while achieving Sequoia compatibility.

### Permission-based workarounds
Multiple projects discovered that **file location and permission fixes** resolve many Sequoia installation issues:
- Moving .pkg files from cloud-synced folders to local directories
- Granting Full Disk Access to the system Installer app
- Installing to external drives as destinations

These solutions could provide immediate relief while longer-term fixes are developed.

## Significant Testing and Validation Gaps

The original analysis lacks **comprehensive validation scenarios** critical for enterprise deployment:

**Multi-Version Compatibility**: No evidence of testing across different Sequoia versions (15.0-15.4) or validation of Intel vs Apple Silicon compatibility differences.

**Enterprise Context**: Missing evaluation of corporate/MDM deployment scenarios where packages behave differently under enterprise security policies.

**Upgrade/Downgrade Scenarios**: No testing of behavior when upgrading from working older installations or when upgrading macOS versions.

**Edge Cases**: Insufficient consideration of case-sensitive filesystems, encrypted disks, proxy configurations, and certificate validation in restricted environments.

## Working Interim Solutions Already Available

Investigation revealed that **command-line binary installation currently works** on Sequoia with known workarounds for post-installation issues:
- Binary installation via `chmod +x meshagent && ./meshagent -install` succeeds
- Post-reboot connection issues can be resolved by renaming daemon labels
- Permission issues are manageable through System Settings

This suggests that **improving documentation and automation of the binary installation process** could provide immediate user relief while more complex solutions are developed.

## Implementation Risk Assessment

The proposed pkgbuild solution introduces **significant risks and costs**:

**Financial Impact**: Requires ongoing Apple Developer Program subscription costs
**Maintenance Overhead**: Substantially increases build system complexity
**User Experience**: Even properly signed packages require user approval in Sequoia
**Development Time**: Major architecture changes to build/release processes

Given that working alternatives exist, the **cost-benefit ratio** of the pkgbuild approach may not justify its implementation complexity.

## Recommended Implementation Strategy

Based on this comprehensive review, a **phased approach** would be more appropriate:

**Phase 1 (Immediate)**: Enhance binary installation documentation and create automated setup scripts to address the current Sequoia compatibility crisis.

**Phase 2 (Short-term)**: Investigate flat package conversion using `pkgutil --flatten` to address format compatibility with minimal code changes.

**Phase 3 (Long-term)**: Evaluate Apple Developer Program membership and notarization infrastructure only if simpler solutions prove insufficient.

## Conclusion 

While the original analysis correctly identified the fundamental compatibility issue, it **underestimated implementation complexity** and **overlooked simpler alternative solutions**. The proposed pkgbuild approach, while technically sound, introduces significant costs and complexity that may be unnecessary given available alternatives.

A more pragmatic approach focusing on immediate user relief through improved binary installation processes, combined with format conversion techniques, would likely deliver better outcomes with substantially lower implementation costs and risks.

**Comment by si458** [comment]
@Junosprite007 thank you for the help/suggestion
unforunately we cant use pkgbuild as what about windows and linux servers running meshcentral?
from my understanding pkgbuild is OSX ONLY
but also more unforunately this issue wont be fixed until https://github.com/Ylianst/MeshCentral/issues/5653 and https://github.com/Ylianst/MeshCentral/issues/6402 are fixed as they stop the app working at all on Sequoia

**Comment by jeremy-bodoni** [comment]
Having spent a good two weeks on similar (but much more painful) issues in our deployment and finally solving them. I wanted to come and update all the projects I've found who are facing similar issues - mpkgs are now legacy and (depending on your OS version) will just be outright rejected by Installer.app. You can just switch to a distribution .pkg using the exact same distribution.xml (or .dist as its named in your files)

Pkgutil --flatten also has issues when making the top level Distribution pkg... productbuild is the only way I've got our multiple component Distribution PKG working. It also wants the internal pkgs directly inside the Distribuition PKG Contents (/MeshAgents.pkg/Internal.pkg instead of /MeshAgents.mpkg/Content/Packages/Internal.pkg)  or it will hit "Unable to use PK session due to incompatible packages. Terminating." this seems to be a bug in the Private Install.Framework on macOS which also prevents Distribution Pkgs with External Components (https and file://) from running.

In you dist you will need to change

`    <pkg-ref id="internal.pkg" installKBytes="###INSTALLSIZE###" version="1.0" auth="Root">file:./Contents/Packages/internal.pkg</pkg-ref>`

to 

`    //<pkg-ref id="internal.pkg" installKBytes="###INSTALLSIZE###" version="1.0" auth="Root">#internal.pkg</pkg-ref>` to reference it once it is moved to /MeshAgents.pkg/Internal.pkg

although 

`    <pkg-ref id="internal.pkg" installKBytes="###INSTALLSIZE###" version="1.0" auth="Root">#Contents/Packages/internal.pkg</pkg-ref>` 

_might work_ with your files if you absolutely need to keep it in the current file structure as when installer.app looks to open an internal package the pre path is set to `/[path_to_parent_pacakge]/`

or productbuild can automatically synthesise the distribution.xml for you if you run it on a folder.

Pkgutil is fine for making the internal component pkgs however so continue using it there!

The pkgutil --expand seems to not work but **it does work** - it is just the actual resulting internal.pkg you see in Finder (after running the expand command) is **a folder** not a pkg - renaming it from internal.pkg to internal.zip (and then removing the .zip) will make finder update and show it as a browsable folder. You can then pkgutil --flatten that folder to turn it back into a pkg.

"Note: When you are inspecting the expanded file structure in Finder, it will show this subdirectory with the package icon. The folder name ends with .pkg which Finder erroneously interprets as a file extension for a bundle-style installation package. If you want to see the contents of this folder in Finder, choose ‘Show Package Contents’ from the context menu." (Source: https://scriptingosx.com/2025/07/inspecting-packages/)

Hope this helps :)

Jeremy

**Comment by mattiacolombomc** [comment]
hey, any updates on this? I agree with @jeremy-bodoni , migrating the deprecated `mpkgs` would be ideal

**Comment by si458** [comment]
@mattiacolombomc sadly nothing yet 👎 
the only thing we have is a branch from a gent who managed to fix the agent, and then built an `.app` which seemed to partly work!
but the down side is YOU (the admin of your meshcentral server) MUST have an apple developer account to code-sign the .app and binarys and preferabily the paid one $99 a year

**Comment by mattiacolombomc** [comment]
is it completely working? with the complete set of features?

could you please link it?

luckily I already have an Apple Dev Account...

for a general release, usually when downloading Github .app on macOS xattr -cr / removing quarantine is needed.

Is there something I'm missing? Is the signing influential in other fields than starting the app?

**Comment by si458** [comment]
@mattiacolombomc well i managed to build the binary and run i without using the `.app`
but the was also the `.app` which again i managed to build and get working
bu only problem i kept having was the permissions kept being removed?
also i couldnt customise the name, i was stuck with `meshagent` when i wanted to use my own app name :(
https://github.com/PeetMcK/MeshAgent/tree/_100-PR-macOS-MeshAgent-Rebuild

**Comment by chrisrgling** [comment]
Hello,
I wanted to try installing MeshCentral on our Macs today, but I’m getting a message saying that the version isn’t compatible with the new version of macOS. I see that you’re still working on this. Let’s hope a solution is found soon.

Have a great day, and best of luck

<img width="656" height="972" alt="Image" src="https://github.com/user-attachments/assets/7b62c892-9e36-48c2-a54b-8048934d05f1" />

<img width="512" height="374" alt="Image" src="https://github.com/user-attachments/assets/6d72b66c-b475-4891-8dee-19ac69d381f0" />

**Comment by si458** [comment]
finally been fixed by a kind gent (or ai not sure?) with the PR https://github.com/Ylianst/MeshCentral/pull/7837
works a treat now!
will need to fix customisation like image and logo etc, but it installs correctly and works on start up
ONLY WORKS ON STARTUP EVERYTIME IF YOU HAVE FILEVAULT DISABLED!
IF YOU HAVE FILEVAULT DISABLED, THE END USER MUST LOGIN FIRST THEN MESHAGENT WILL RUN!

**Comment by johnnyq** [comment]
@si458 This is fantastic news!

-now last up is Linux Wayland support and this thing is on blast off again!!

**Comment by strange-qwq** [comment]
Has this issue been fixed yet? When I use the 1.2.0 server to download the macOS mesh agent, I still get an installation failure message.

**Comment by si458** [comment]
@strange-qwq no I'm hoping in 1.2.1 it will be fixed! Fingers crossed!

**Comment by ariesadonis** [comment]
Hi, saw this installation issue. I'm building a tool that scans a repo's dependency and env files to generate an onboarding/setup doc automatically, so version-specific install problems like this are easier to spot and document. Happy to run it against MeshCentral and share the output if that would help.

**Comment by DavidGarciaCat** [comment]
Hello, @si458 

First of all, thanks for referring from the other issue (https://github.com/Ylianst/MeshCentral/issues/7921) to this one; it may be better if we unify the communication, as they seem related, despite the 1.5 years since this issue was opened.

I am not a C or C++ expert (as I have read a few comments above that someone with more experience could take a look), but unless I am mistaken, new macOS ARM/Intel/universal binaries have been compiled and committed to master in [5814c37](https://github.com/Ylianst/MeshCentral/commit/5814c372889a0e65ad9b337678232841906f8c02). However, despite getting the new binaries (that should fix the issue we have), no new releases (maybe v1.2.1) have been tagged and made publicly available (as we are still using v1.2.0).

Thus, I would like to confirm what the next steps are, and maybe consider an action plan to avoid additional delays for this fix, as other users may face the same issue - and they could just not report it at all.

Please advise, and thanks,

**Comment by si458** [comment]
@DavidGarciaCat wait for 1.2.1 release, then upgrade ur server, then redownload the meshagent for mac and give it a try!

**Comment by DavidGarciaCat** [comment]
> [@DavidGarciaCat](https://github.com/DavidGarciaCat) wait for 1.2.1 release, then upgrade ur server, then redownload the meshagent for mac and give it a try!

Yeah, that's the usual plan - but would it be possible to narrow a little bit an ETA for this? Just wondering

**Comment by si458** [comment]
@DavidGarciaCat normally once a month and next one due thursday 🙃

**Comment by DavidGarciaCat** [comment]
> [@DavidGarciaCat](https://github.com/DavidGarciaCat) normally once a month and next one due thursday 🙃

👍 much appreciated for your confirmation - looking forward to getting the new fix!

**Comment by si458** [comment]
Plz upgrade to 1.2.1 and download the new pkg and give it a try!
Should work a treat now!

**Comment by cvoisey** [comment]
The good news is that I see the screen now, but it only shows the background and the top menu bar.  I can not type and I can not see anything but the screen background.  Windows and contents are not shown.  

I have looked at my Privacy & Security and even toggled some of these, including Screen Recording off and on again.

**Comment by si458** [comment]
@cvoisey make sure u do it in order.

1. Uninstall old meshagent
2. Remove permissions from screen recording, full file access and accessibility
3. Restart mac
4. Download NEW meshagent from web ui (1.2.1 release is needed) and make sure u use the correct version, DONT USE UNIVERSAL!!! USE SILICON OR INTEL only!
5. Install meshagent
6. Give it permission for full file access
7. Try connect to mac, accept the screen recording
8. Then u might be able to click accessibility and accept mouse/keyboard control but if meshagent isn't listed, just disconnect from the mac and connect again and u should get another prompt to accept

Sadly the is a very weird bug with permissions where if the permissions are already there from previously, it wont work until u remove them and restart

**Comment by cvoisey** [comment]
I did manage to get the screen to eventually show up, probably using the ARM download vs. Universal.  That said, I am not able to control input and when the computer is restarted, the service does not restart itself.  It has to be installed each time.  

This issue should remain open (reopened) until we get all this completed fully.  I appreciate all the work so far, and hopefully we can get the finally amount soon!

**Comment by si458** [comment]
@cvoisey I've tested it on 4 different macs and version and it works on all of them so I'm not sure what's wrong?
If u install it then run `info` on the console tab of the web ui, what version/core date does it say u have installed?

**Comment by cvoisey** [comment]
I've gone through 3 different Macs so far, 2 on Apple Silicon and 1 Intel.  All same results.  

Current Core: Jun 2 2026, 3389938298
Agent Time: 2026-06-24 15:48:44.935-07:00.
User Rights: 0xffffffff.
Platform: darwin.
Capabilities: 15.
Server URL: wss://<redacted>:443/agent.ashx.
OS: macOS 26.5.1.
Modules: computer-identifiers, linux-dhcp, linux-info, mac-info, monitor-border, sysinfo, util-agentlog, wifi-scanner-windows, wifi-scanner.
Server Connection: true, State: 1.
Node ID: <redacted>
Application Location: /usr/local/mesh_services/meshagent/meshagent/

**Comment by strange-qwq** [comment]
@cvoisey I'm also on macOS 26.5.1, but I didn't run into the problem you mentioned after installation. The only hiccup was that the Accessibility permission prompt didn't appear when I first tried to control — it only popped up after I repeatedly refreshed the page and reconnected to the desktop. You can try checking **Privacy & Security** → **Accessibility** and **Screen & System Audio Recording** to see if the **meshagent** app is listed there and whether its toggle is enabled. Once I turned both on, everything worked perfectly. If it still doesn't work after enabling them, try restarting your system. If the issue persists after a reboot, please open a separate issue — this one was specifically about fixing the installation failure, and control issues are outside its scope.

**Comment by MCal27** [comment]
The update list looks amazing! especially for macOS. sadly no matter what I try I'm getting the old 'installation failed' error when trying to install the agent :( I unstalled the old agent using the script and when that didnt seem to get rid of it, I removed the meshagent folder in Library. I also removed all entries in Accessability, screen recording and full disk access for the meshagent too, and rebooted after. What am I missing folks?!

**Comment by cvoisey** [comment]
I appreciate the work that went into 1.2.1, and it does look like progress was made.  However, I don’t think this issue should be closed yet.

From my side, I have now tested this on 3 different Macs, 2 Apple Silicon and 1 Intel, and I am still not seeing a fully working result.  The screen eventually showed up after using the architecture-specific installer, but input control is still not working, and the agent does not survive reboot properly.  It has to be installed again.

I understand the original symptom was “installation failed,” but the practical outcome here needs to be that the macOS agent installs cleanly, launches correctly, survives reboot, shows the desktop, and allows remote control after the required permissions are granted.  Until that is verified across affected macOS 15/26 systems, I don’t think this should be treated as completed.

I have developed and run projects for decades, including much larger ones than this, and I would not close a task like this until the fix is verified end-to-end.  Right now it feels like a fix was added, the issue was marked done, and the remaining validation is being pushed back onto users.  That creates exposure because people may assume this is resolved when it is not fully working yet.  We even have another user above reporting another issue.

I’m happy to keep testing and provide logs or specific steps, but I think this should be reopened until the install, permissions, control, and reboot behavior are confirmed and we have enough users.  If this is a side case, we can absolutely go that direction.  But there isn't enough info to declare that at this time, based on what I see above.  The only clear test is the developer, si458.

**Comment by si458** [comment]
@MCal27 @cvoisey strange question, whats the OS you are using to host meshcentral?
a gent who sent us his pkg for me to look at says he uses Windows
so im wondering if its a windows server bug generating the pkg?
as i also did some debugging and you can extract the pkg and install it manually without issues
so its just the installer wizard that seems broken?

**Comment by MCal27** [comment]
@si458 I'm hosting on Windows server 2016. I'll look into the .pkg tomorrow. thanks

**Comment by si458** [comment]
@MCal27 u have answered my question!!!
Ive also replicates it here, for some reason it wont build correctly on windows?
If u host meshcentral with linux, works no problems...
So back to the drawing board for the moment (if ur a windows hosted that is)

**Comment by cvoisey** [comment]
For me, MeshCentral is hosted on an Ubuntu box on AWS. 

I did a bit more troubleshooting and found the following:
- got it working on my Intel-based macOS Sequoia 15.2.  All issues above with process, etc. were seen, but it eventually worked as prescribed.
- identical machine, but running Tahoe 26.5.1 gets installed and I can see the screen, but I I can not control it and it doesn't load on reboot
- Apple Silicon-based macOS Tahoe 26.5.1, same result as above.  installs and eventually managed to get screen to show, but unable to control input.  It appears that the Accessibility permissions are never triggered.  On reboot, I do see that meshagent process is running, but it does not connect to the Mesh Central server.  Something looks blocked.
- Same configuration as previous, but only Tahoe 26.5.  All results are identical. 
- Curiosity-wise, I just tried it on my dev machine for Golden Gate DB2 and while I don't expect any real support on a developer beta, it appears that the ability to see the screen is not working. There could be a new layer in their permissions to watch out for.  We haven't seen this in any of our apps or ones we use, but I figured it may be worth throwing out there.  

For the newer builds on Tahoe and now Golden Gate, I know Apple has been hitting hard on the network permissions.  Do you have to add a permission for MeshAgent to use the Local Network now?  That could explain why it doesn't come back on boot.

**Comment by cvoisey** [comment]
I did some fiddling .... 

It is permissions issues, likely started in some specific build of macOS.  The fixes are easy, and maybe you can incorporate them somehow.  

In order to regain control and have meshagent start again, you just need to change the permissions and relaunch:
sudo rm -f /Library/LaunchAgents/*mesh*.plist
sudo chown root:wheel /Library/LaunchDaemons/*mesh*.plist
sudo launchctl bootstrap system /Library/LaunchDaemons/meshagent.plist

For macOS 27 GG DB2, the additional fix is to drag and drop your meshagent from /usr/local/mesh_services/meshagent/meshagent onto the "Screen & System Audio Recording" permissions list.  It seems this does not trigger on its own in this release, at least.

**Comment by MCal27** [comment]
> [@MCal27](https://github.com/MCal27) u have answered my question!!! Ive also replicates it here, for some reason it wont build correctly on windows? If u host meshcentral with linux, works no problems... So back to the drawing board for the moment (if ur a windows hosted that is)

Thanks for the prompt reply. My host is pretty well configured as I need it. if possible (and I'm not anti linux by any means) I'd prefer to keep it on Windows if at all possible. Thanks for your work :)

**Comment by cvoisey** [comment]
FYI, there's another bug in this.  While the above does work, it appears the Screen & System Audio Recording work once and then the next time you have to toggle it.  This is unique to MeshCentral, so it is something within this code.

**Comment by si458** [comment]
@cvoisey I think u are chasing bugs that's already been fixed???

All ur success work arounds you list above are what we already do in in the new meshagent and the pkg file?

Are u sure u updated to 1.2.1 and then redownloaded the new plg arm64 version or intel version? (Dont use universal!)

You can check the macosinstaller.js file in the repo and see how we build the pkg with all these options in?

The problem at the moment is ubuntu servers hosting meshcentral work building the pkg correctly but it seems windows server hosting meshcentral dont build the pkg correctly? (I have verified this myself just before and im going to look into it next week properly find out why)

**Comment by cvoisey** [comment]
You are absolutely right to question that one!  Inspecting the installed binary, I see it is has itself labelled as 1.2.11  

I am not downloading the arm version and I did try it both through the link and direct distribution.  This wouldn't be the first time I have seen different version for different distributions.  But that doesn't seem to be the case.  

I am also NOT downloading the Universal versions.  100% sure of that.

**Comment by stanelie** [comment]
So,

I made a working solution to this issue with Claude.

The main issue was that the login window cannot be interacted with by any other mean than a real physical keyboard, or macOS's own screen sharing. So, this new agent has 2 agents, one for the pre-login screen, and one for the user session. The pre-login one uses screen-sharing to send keystrokes and mouse events (so you can actually login) to the mac VNC server, on a private loopback connection, using a install-generated unique random password. I think this is secure.

I put everything on a github repo and the idea is that I hope this will demonstrate how to do the whole thing so that the developers of Meshcentral can do it properly. My version is clunky and requires some manual steps.

The repo is https://github.com/stanelie/meshcentral-mac-agent

This has all been written by Claude Code, even the documentation, so, I won't be able to support anyone trying to implement this. I was just there to test and guide Claude, all the code is AI written.

But it works for me, and it works great!

I hope this helps.

**Comment by si458** [comment]
@stanelie looks good, but issue, you cant control the lock screen on very first start up when you have filevault enabled
not even teamview or anydesk or rustdesk have this support
you might want to make a note saying that it wont work with filevault enabled!

**Comment by stanelie** [comment]
This is correct : with filevault enabled, the first login prompt actually happens before the mac is booted, and so, meshagent does not run yet, nothing runs yet.

**Comment by stanelie** [comment]
I updated the documentation so it is clear that filevault needs to be off.
